Add ExtendRestAPI and SlotFills docs (#4208)
woocommerce/woocommerce-gutenberg-products-block: docs/extensibility/extend-rest-api-add-data.md

> # Exposing your data in Store API.

Reviewer: ```suggestion
# Exposing your data in the Store API.
```

## The problem
You want to extend the Cart and Checkout blocks, but you want to use some custom data not available on Store API or the context.
You don't want to create your own endpoints or Ajax actions. You want to piggyback on the existing StoreAPI calls.

## Solution
ExtendRestAPI offers the possibility to add contextual custom data to Store API endpoints, like `wc/store/cart` and `wc/store/cart/items` endpoints.

> That data is namespaced to your plugin and protected from other plugins malfunctioning.

Reviewer 2: ```suggestion
That data is namespaced to your plugin and protected from other plugins causing it to malfunction.
```

The data is available on all frontend filters and slotFills for you to consume.

## Basic usage

> You can use ExtendRestAPI by registering a couple of functions, `schema_callback` and `data_callback` on a specific endpoint namespace. Those functions are going to be called at execution time and are going to be passed relevant data.

Reviewer 2: The phrasing here feels a bit difficult for me to follow.  Maybe something like this:

```suggestion
You can use ExtendRestAPI by registering a couple of functions, `schema_callback` and `data_callback` on a specific endpoint namespace. ExtendRestAPI will call them at execution time and will pass them relevant data as well.
```

```PHP

use Automattic\WooCommerce\Blocks\Package;
use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartSchema;

add_action('woocommerce_blocks_loaded', function() {

>  // ExtendRestApi relies on a Singleton that is shared between the API and consumers.

Reviewer 2: These two comments about using the singleton of ExtendRestAPI are really helpful.  It might be worth including them in some of the higher-level guidance above.  Maybe there could be a section along the lines of "considerations" or "important things to keep in mind" or even a step-by-step overview of what each important bit of code in the example will accomplish.

Author: That's a great idea!

Author: cc @nerrad because my PHP and DI knowledge is rusty.

Is it possible for us to have `new ExtendRestApi` still fetch the same instance as `Package::container()->get( ExtendRestApi::class )`?
I think I saw patterns like this where `ExtendRestApi` constructor would just assign its return value to whatever is in `Package::container()->get( ExtendRestApi::class )`.

This would require us to change two things:
- Change ExtendRestApi so it's initiated with `init` function, calling `__construtor` would just return whatever `Package::container()->get( ExtendRestApi::class )`.
```php
public function __construct() {
	return Package::container()->get( self::class );
}

public function init( Package $package, Formatters $formatters ) {
	$this->package    = $package;
	$this->formatters = $formatters;
}
```
- This code block in `Domain/Bootstrap.php` would accommodate that change somehow
```php
$this->container->register(
	ExtendRestApi::class,
	function( Container $container ) {
		return new ExtendRestApi( $container->get( Package::class ), $container->get( Formatters::class ) );
	}
);
```

Reviewer 3: > Is it possible for us to have new ExtendRestApi still fetch the same instance as Package::container()->get( ExtendRestApi::class )?

Constructors don't return their value, their purpose is to instantiate the class. 

So this won't work the way you expect.  There's the typical singleton pattern whereby a class has an `static instance()` method to return the cached instance of the class from the class itself, but singletons have their own problems in PHP (especially because of their overuse). 

Besides that, I don't think I would technically call this a Singleton (although we treat it like one). It's more of a shared instance because the instance is cached in the container. 

```suggestion
 // ExtendRestApi is stored in the container as a shared instance between the API and consumers.
```

>  // You shouldn't initiate your own ExtendRestApi using new ExtendRestApi but should always get it from Package dependecy injection manager.

Reviewer 3: ```suggestion
 // You shouldn't initiate your own ExtendRestApi instance using `new ExtendRestApi` but should always use the shared instance from the Package dependency injection container.
```

>  $extend = Package::container()->get( ExtendRestApi::class );

Reviewer 3: In your example, I think it's important that `Package::container()` is only used when the Blocks plugin is loaded, otherwise there will be fatal errors. This action is a hookpoint intended especially for that and should be used in the example: 

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/d226576b087296bdf1039a6aa790469ab0863c73/src/Domain/Bootstrap.php#L64

So you could wrap the implementation here in a callback registered to that action.

Reviewer 3: **Edit:** I see you do this in the full example. There is an explanation accompanying it, but I wonder if it's something folks could miss 🤔 (This doesn't necessarily mean you need to modify/improve it).

Author: On the same example (not the full) I do wrap it in an action callback, see line 21 of this file.

Reviewer 3: Doh 🤦🏻 I missed that! Awesome 👍🏻

 $extend->register_endpoint_data(
 array(
 'endpoint' => CartSchema::IDENTIFIER,
 'namespace' => 'plugin_namespace',
 'data_callback' => 'my_data_callback',
 'schema_callback' => 'my_schema_callback',
 'schema_type' => ARRAY_A,
 )
 );
});

function my_data_callback() {
 return [
 'custom-key' => 'custom-value';
 ]
}

function my_schema_callback() {
 return [
 'custom-key' => [
 'description' => __( 'My custom data', 'plugin-namespace' ),
 'type' => 'string'
 'readonly' => true,
 ]
 ]
}
```

> Data callback and Schema callback can also receive parameters:

Reviewer 2: I saw that you document these parameters over on a separate file (which I think is fine), but it might be worth providing a link to them either here or in the section above so that folks can jump over and find out which endpoints provide which data to each callback type.

```PHP

function my_cart_item_callback( $cart_item ) {
 $product = $cart_item['data'];
 if ( is_my_custom_product_type( $product ) ) {
 $custom_value = get_custom_value( $product );
 return [
 'custom-key' => $custom_value;
 ]
 }
}

```

## API Definition

- `ExtendRestAPI::register_endpoint_data`: Used to register data to a custom endpoint. It takes an array of arguments:

| Attribute | Type | Required | Description |
| :-------- | :----- | :------: | :------------------------------------ |

> | `endpoint` | string | Yes | The endpoint you're trying to extend, it is suggested that you use `::IDENTIFIER` Available on the route Schema class to avoid typos. |

Reviewer: ```suggestion
| `endpoint` | string | Yes | The endpoint you're trying to extend. It is suggested that you use the `::IDENTIFIER` available on the route Schema class to avoid typos. |
```

| `namespace` | string | Yes | Your plugin namespace, the data will be available under this namespace in the StoreAPI response. |
| `data_callback` | callback | Yes | A callback that returns an array with your data. |
| `schema_callback` | callback | Yes | A callback that returns the shape of your data. |

> | `data_type` | string | No (default: `ARRAY_A` ) | The type of your data, if you're adding an object (key => values), it should be `ARRAY_A`, you're adding a list of items, it should be `ARRAY_N`. |

Reviewer: ```suggestion
| `data_type` | string | No (default: `ARRAY_A` ) | The type of your data. If you're adding an object (key => values), it should be `ARRAY_A`. If you're adding a list of items, it should be `ARRAY_N`. |
```

> | `data_type` | string | No (default: `ARRAY_A` ) | The type of your data. If you're adding an object (key => values), it should be `ARRAY_A`. If you're adding a list of items, it should be `ARRAY_N`. |

Reviewer 2: Do the letters `A` and `N` have a certain significance here?  It might be worth including that to help developers remember which is which.

Author: ARRAY_A means associative array (key => value), ARRAY_N means numeric array (0 => value, 1 => value).

https://github.com/WordPress/WordPress/blob/270f2011f8ec7265c3f4ddce39c77ef5b496ed1c/wp-includes/wp-db.php#L32

They're commonly used in WordPress when fetching data from the database
and deciding the shape of returned data.

If you feel like they don't make sense (I kinda do), I'm open to another suggestion.

This was recently added.

## Putting it all together

> This is a complete example that shows how you can register contextual WooCommerce Subscriptions data in each cart item (simplified).

Reviewer 2: It's great that you have included a complete example here! :sparkles: 🙌

> This example uses Formatters, another property documented here (TBD).

Author: Created follow up issue for Formatters https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4210

Reviewer: I'd suggest maybe adding a note to that issue that the information should be cross-linked here once it's in place.

```php
<?php
/**
 * WooCommerce Subscriptions Extend Store API.
 *
 * A class to extend the store public API with subscription related data
 * for each subscription item
 *
 * @package WooCommerce Subscriptions
 */

use Automattic\WooCommerce\Blocks\Package;
use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
use Automattic\WooCommerce\Blocks\StoreApi\SchemaController;
use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartItemSchema;

if ( class_exists( 'Package' ) && version_compare( Package::get_version(), '4.8.0', '>=' ) ) {
	// This class needs to run after WooCommerce Blocks is ready.
	add_action( 'woocommerce_blocks_loaded', array( 'WC_Subscriptions_Extend_Store_Endpoint', 'init' ) );
}

class WC_Subscriptions_Extend_Store_Endpoint {

	/**
	 * Stores Rest Schema Controller.
	 *
	 * @var ExtendRestApi
	 */
	private static $schema;

	/**
	 * Stores Rest Extending instance.
	 *
	 * @var ExtendRestApi
	 */
	private static $extend;

	/**
	 * Plugin Identifier, unique to each plugin.
	 *
	 * @var string
	 */
	const IDENTIFIER = 'subscriptions';

	/**
	 * Bootstraps the class and hooks required data.
	 *
	 * @since WCBLOCKS-DEV
	 */

> 	public static function init() {
> 		self::$schema = Package::container()->get( SchemaController::class );
> 		self::$extend = Package::container()->get( ExtendRestApi::class );
> 		self::extend_store();
> 	}

Reviewer 3: Nit, to promote DI patterns a bit more, it'd be good to pass these in on the `init` method rather than instantiating in the class:

```suggestion
	public static function init( SchemaController $schema_controller, ExtendRestAPI $extend_rest_api ) {
		self::$schema = $schema_controller;
		self::$extend = $extend_rest_api;
		self::extend_store();
	}
```

There's at least few reasons for this:

- Direct container access should typically happen in as few places as possible (typically in bootstrapping) because usage elsewhere is typically bad usage of the service locator pattern (which we have here).
- Unit Testing this would be much more difficult. While it'd be possible to swap `SchemaController` and `ExtendRestAPI` mocks into the container, it's not trivial and requires understanding how the container works. Instead, with passing them in as dependencies, they could be mocked much more easily for testing.
- The class is tightly coupled to the availability of `Package::container()`. While your example does make sure it's available _outside this class_ by only initializing it on the `woocommerce_blocks_loaded` action, note _that happens outside the class_. Let's make that more explicit and also pass in the SchemaController and ExtendRestAPI _outside the class_ in the same logic also testing for the availability.
